enderio-machines/src/main/java/com/enderio/machines/common/blockentity/AbstractMachineBlockEntity.java

```diff
@@ -0,0 +1,181 @@
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import net.minecraftforge.items.CapabilityItemHandler;
+import net.minecraftforge.items.IItemHandler;
+
+import static net.minecraftforge.fluids.capability.IFluidHandler.FluidAction;
+
+import java.util.EnumMap;
+import java.util.Optional;
+
+public abstract class AbstractMachineBlockEntity extends SyncedBlockEntity implements MenuProvider {
+
+    private final IOConfig config = new IOConfig();
+
+    @Getter
+    @Setter
+    private RedstoneControl redstoneControl = RedstoneControl.ALWAYS_ACTIVE;
+
+    private final EnumMap<Direction, LazyOptional<IItemHandler>> itemHandlerCache = new EnumMap<>(Direction.class);
+    private final EnumMap<Direction, LazyOptional<IFluidHandler>> fluidHandlerCache = new EnumMap<>(Direction.class);
+    private boolean isCacheDirty = false;
+
+    public AbstractMachineBlockEntity(BlockEntityType<?> pType, BlockPos pWorldPosition, BlockState pBlockState) {
+        super(pType, pWorldPosition, pBlockState);
+        add2WayDataSlot(new EnumDataSlot<>(this::getRedstoneControl, this::setRedstoneControl, SyncMode.GUI));
+        add2WayDataSlot(new NBTSerializableDataSlot<>(() -> config, SyncMode.RENDER));
+    }
+
+    public final IOConfig getConfig() {
+        return this.config;
+    }
+
+    public abstract ItemHandlerMaster getItemHandlerMaster();
+
+    public void updateCache() {
+        itemHandlerCache.clear();
+        fluidHandlerCache.clear();
+        for (Direction direction: Direction.values()) {
+            BlockEntity neighbor = level.getBlockEntity(worldPosition.relative(direction));
+            if (neighbor != null) {
+                itemHandlerCache.put(direction, addInvalidationListener(neighbor.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, direction.getOpposite())));
+                fluidHandlerCache.put(direction, addInvalidationListener(neighbor.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, direction.getOpposite())));
+            } else {
+                itemHandlerCache.put(direction, LazyOptional.empty());
+                fluidHandlerCache.put(direction, LazyOptional.empty());
+            }
+        }
+    }
+
+    /**
+     * needs to be called to prevent an instant call of the listener if the capability is not present
+     * @param capability
+     * @param <T>
+     * @return
+     */
+    private <T> LazyOptional<T> addInvalidationListener(LazyOptional<T> capability) {
+        if (capability.isPresent())
+            capability.addListener(this::markCacheDirty);
+        return capability;
+    }
+    private <T> void markCacheDirty(LazyOptional<T> capability) {
+        isCacheDirty = true;
+    }
+
+
+    public static void tick(Level pLevel, BlockPos pPos, BlockState pState, AbstractMachineBlockEntity pBlockEntity) {
+        pBlockEntity.tick();
+    }
```

Comment:
A few spacing things like this in this file, some too many newlines and some too few

Reply:
Is this how you like it? I've added a few more newlines to https://github.com/Rover656/EnderIO-Rewrite/pull/29/commits/9a5b1eb40be78324237f1338fbdff152e6812363 ,but I'm unsure if that's how you intended it to be

Reply:
yup; looking good, just noticed there were some methods without spacing and some with doubles was all :+1: